### PR TITLE
Update Cwd.pm to work on Guix & NixOS

### PR DESCRIPTION
**Pull request by Frosku**
refs #18703 

This change fixes Cwd on NixOS/Guix.

### CODE REVIEW

**Comment by grahamc** [comment]
The more correct way to handle this (as a NixOS contributor) is to patch Cwd at build time to use an absolute path to pwd. Please don't merge this.

**Comment by Frosku** [comment]
> The more correct way to handle this (as a NixOS contributor) is to patch Cwd at build time to use an absolute path to pwd. Please don't merge this.

As discussed in IRC, this doesn't work when using any of the common perl version managers (perlbrew/plenv) which do not use nix expressions or nix build tools.

**Comment by grahamc** [comment]
Those tools should also be patched.

**Comment by Frosku** [comment]
> Those tools should also be patched.

This would require actual rewriting of the tool so that it rewrites `Cwd.pm` in place every time it pulls perl.

**Comment by grahamc** [comment]
Unfortunately this will be true for lots of software installed by perlbrew etc, as lots of Perl packages expect programs at specific places in the FS. I wonder what other techniques might work?

**Comment by Frosku** [comment]
> Unfortunately this will be true for lots of software installed by perlbrew etc, as lots of Perl packages expect programs at specific places in the FS. I wonder what other techniques might work?

Typically this is for the same reason where you see it: to avoid breaking under [taint mode](https://perldoc.perl.org/perlsec#Taint-mode), which is an ancient security feature. Short of deprecating taint mode or rewriting perl version managers (plenv/perlbrew) to rewrite the module in place before building, it's very difficult to solve.

Regardless, perl should be able to be built in nixos with or without using nix build tools, because current nix build tools do not have the sophistication to replace something like perlbrew (managing multiple versions, and multiple `local::lib` setups). Tim Toady and all that.

**Comment by jkeenan** [comment]
@tonycoz or @atoomic, could either of you review the discussion in this p.r.?

Thank you very much.
Jim Keenan

File: dist/PathTools/Cwd.pm [hunk 1]
```diff
@@ -155,6 +155,8 @@ if($^O ne 'MSWin32') {
     foreach my $try ('/bin/pwd',
 		     '/usr/bin/pwd',
 		     '/QOpenSys/bin/pwd', # OS/400 PASE.
+		     '/run/current-system/profile/bin/pwd', # GUIX
+		     '/run/current-system/sw/bin/pwd' # NixOS
```
Comment (grahamc) [inline dist/PathTools/Cwd.pm:159]: Importantly, Nix's concept of closures is very important and highly valued. It is very common to export a closure and run it elsewhere, where `/run/` doesn't exist in any form. This works "sometimes" but intentionally sidesteps the design philosophy of Nix, creating software artifacts that appear to work in development and then fails on deployment due to cheating like this.

**Comment by tonycoz** [comment]
I suspect from the discussion and from using nixos to test #18791 that Configure should be detecting pwd (or PATH) to save in `config.sh` and Cwd should then using that to find pwd by default.

I think the standard way would be to use confstr() to get the system PATH and search that, but POSIX.pm doesn't expose that, and POSIX.pm isn't available at the point this is called anyway.  I'm pretty sure that would have the same problems that this change does, since the result of `getconf PATH` doesn't change in a child environment.

So from the discussion I don't think this change is appropriate.

**Comment by jkeenan** [comment]
> I suspect from the discussion and from using nixos to test #18791 that Configure should be detecting pwd (or PATH) to save in `config.sh` and Cwd should then using that to find pwd by default.
> 
> I think the standard way would be to use confstr() to get the system PATH and search that, but POSIX.pm doesn't expose that, and POSIX.pm isn't available at the point this is called anyway. I'm pretty sure that would have the same problems that this change does, since the result of `getconf PATH` doesn't change in a child environment.
> 
> So from the discussion I don't think this change is appropriate.

No one has disputed @tonycoz's judgment on this question in the past five months, so I'm closing this p.r.

**Comment by Leont** [comment]
> I suspect from the discussion and from using nixos to test #18791 that Configure should be detecting pwd (or PATH) to save in config.sh and Cwd should then using that to find pwd by default.

This sounds sensible to me

**Pull request closed without merging**